### PR TITLE
fix(ftplugin): respect runtimepath ordering

### LINKED ISSUE #24003: #23801 breaks slimv
### Problem

It appears that the semantics of plugin loading changed with #23801 -- it looks like before, it'd go thru each directory in rtp, and load each file matching any of the globs. After #23801, it goes one by one thru the globs, then thru each directory, and loads them.

Neovim ships `ftplugin/lisp.vim`, which does the standard

```viml
if exists("b:did_ftplugin")
  finish
endif
let b:did_ftplugin = 1
```

dance. [slimv](https://github.com/kovisoft/slimv/) ships its files as `ftplugin/lisp/*.vim`, and also does the same thing.

Previously, since slimv would be earlier in the rtp search order, it would "win the race" and be loaded. Now, the anemic lisp.vim does instead.

Reverting #23801 fixes this symptom.

### Steps to reproduce

Install slimv, open a Lisp file, type `,c`, note how nothing happens.

### Expected behavior

slimv should be loaded before Neovim's lisp.vim, and should work

### Neovim version (nvim -v)

v0.10.0-dev commit a7e5d4238a00d5bfa5809b2860047eca2d565e62

### Vim (not Nvim) behaves the same?

no

### Operating system/version

NixOS 23.05

### Terminal name/version

kitty 0.28.1

### $TERM environment variable

tmux-256color

### Installation

Git or nixpkgs, tried both

### PR DESCRIPTION
**Pull request by clason**
Problem: bundled `ftplugin/foo/*.vim` are sourced before user
`ftplugin/foo.vim`.

Solution: call `runtime!` once on all patterns to be sourced.

Followup to #23801. Fixes #24003.

### CODE REVIEW

**Comment by clason** [comment]
@gbrlsnchs would you mind checking whether this PR introduces a regression for #16928?

File: runtime/ftplugin.vim [hunk 1]
```diff
@@ -30,9 +30,10 @@ augroup filetypeplugin
       for name in split(s, '\.')
         " Load Lua ftplugins after Vim ftplugins _per directory_
         " TODO(clason): use nvim__get_runtime when supports globs and modeline
-        exe printf('runtime! ftplugin/%s.vim ftplugin/%s.lua', name, name)
-        exe printf('runtime! ftplugin/%s_*.vim ftplugin/%s_*.lua', name, name)
-        exe printf('runtime! ftplugin/%s/*.vim ftplugin/%s/*.lua', name, name)
+        exe printf('runtime! ftplugin/%s.vim ftplugin/%s.lua
+                           \ ftplugin/%s_*.vim ftplugin/%s_*.lua
+                           \ ftplugin/%s/*.vim ftplugin/%s/*.lua',
+                           \ name, name, name, name, name, name)
```
Comment (lewis6991) [inline runtime/ftplugin.vim:36]: Shouldn't this be?

```suggestion
        exe printf('runtime! ftplugin/%s.vim ftplugin/%s_*.vim ftplugin/%s/*.vim
                           \ ftplugin/%s.lua ftplugin/%s_*.lua ftplugin/%s/*.lua
                           \ name, name, name, name, name, name)
```

Comment (clason) [reply runtime/ftplugin.vim:36]: Debatable; would have to test.

Comment (lewis6991) [reply runtime/ftplugin.vim:36]: Otherwise I don't think this PR does anything.

Comment (clason) [reply runtime/ftplugin.vim:36]: Well, it does _something_ (right), according to the person who tested it.

Comment (lewis6991) [reply runtime/ftplugin.vim:36]: Yeah, sorry, I realize now. What I said before wasn't correct.

**Review by zeertzjq — APPROVED** [review]

**Comment by lewis6991** [comment]
Is it worth adding a test so we don't continue to flip-flip on this?

**Comment by zeertzjq** [comment]
I've also noticed that the same ordering problem between .vim and .lua files also applies to every other type of runtime files.

File: test/functional/lua/runtime_spec.lua [hunk 2]
```diff
@@ -138,6 +138,28 @@ describe('runtime:', function()
       exec [[set filetype=new-ft]]
       eq(1, eval('b:lua_ftplugin'))
     end)
+
+    it("'rtp' order is respected", function()
```
Comment (lewis6991) [inline test/functional/lua/runtime_spec.lua:142]: Can you add a comment saying exactly what the expected behaviour should be, or link to the docs or something?

Comment (zeertzjq) [reply test/functional/lua/runtime_spec.lua:142]: Done

**Comment by zeertzjq** [comment]
I'm now thinking about reverting #23805 instead of also backporting this PR to `release-0.9`. These two PRs combined may still be considered a breaking change.

**Comment by lewis6991** [comment]
What's the reasoning? This seems preferable over pre-#23805

**Comment by clason** [comment]
> I'm now thinking about reverting #23805 instead of also backporting this PR to `release-0.9`. These two PRs combined may still be considered a breaking change. Maybe it better to leave ordering in v0.9.x unchanged and settle all runtime file ordering problems (ftplugin, syntax, spell, indent, compiler, colorscheme) in v0.10.

I don't agree; the PR fixed an issue (albeit by adding an unfortunate regression), so reverting just introduces another bug (again). We should keep adding fixes (especially such small fixes) until we get it right.

**Comment by clason** [comment]
> I've also noticed that the same ordering problem between .vim and .lua files also applies to every other type of runtime files.

Yes, as I pointed out elsewhere. It's just that nobody has reported issues for those (and Lua syntax files are unicorns), so I didn't want to touch these until we can switch to `nvim_get_runtime` API for this.

**Comment by clason** [comment]
> Is it worth adding a test so we don't continue to flip-flip on this?

Would be great, if someone could tell me how to test this...

**Comment by zeertzjq** [comment]
I've already added a test

**Comment by zeertzjq** [comment]
There is currently a bug https://github.com/vim/vim/pull/12617 in `source_runtime()` with loading multiple patterns in 'packpath', so the loading order can't currently be fixed for `:colorscheme`. (It's rare to want to override a colorscheme though.) Other types of runtime files only use 2 patterns (one for .vim and one for .lua), unlike ftplugin which uses 6 patterns, so I think it's better to fix other runtime files in a later PR.

**Pull request merged**